e2etest/bgptests/bgp.go

```diff
@@ -147,6 +147,117 @@ var _ = ginkgo.Describe("BGP", func() {
 			}),
 	)
 
+	ginkgo.Describe("GracefulRestart, when speakers restart", func() {
+
+		ginkgo.AfterEach(func() {
+			for _, c := range FRRContainers {
+				c.NeighborConfig.GracefulRestart = false
+			}
+		})
+
+		assertDuringSpeakerRestart := func(grEnabled bool, pairingIPFamily ipfamily.Family, poolAddresses []string, tweak testservice.Tweak) {
+			_, svc := setupBGPService(cs, testNamespace, pairingIPFamily, poolAddresses,
+				FRRContainers, func(svc *corev1.Service) {
+					testservice.TrafficPolicyCluster(svc)
+					tweak(svc)
+				})
+			defer testservice.Delete(cs, svc)
+
+			allNodes, err := cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			testservice.ValidateDesiredLB(svc)
+
+			for _, c := range FRRContainers {
+				validateService(svc, allNodes.Items, c)
+			}
+			err = metallb.RestartSpeakerPods(cs)
+			Expect(err).NotTo(HaveOccurred())
+
+			if !grEnabled {
+				Eventually(func() error {
+					for _, c := range FRRContainers {
+						err := validateServiceNoWait(svc, allNodes.Items, c)
+						if errors.Is(err, ErrStaleRoute) {
+							Expect(err).NotTo(HaveOccurred(),
+								"a stale route cannot be observed if GR disabled")
+						}
+						if err != nil {
+							return err
+						}
+					}
+					return nil
+				}, 2*time.Minute, time.Second).Should(HaveOccurred(), "a downtime should be observed")
+				return
+			}
+
+			Eventually(func() error {
+				for _, c := range FRRContainers {
+					err := validateServiceNoWait(svc, allNodes.Items, c)
+					if errors.Is(err, ErrStaleRoute) {
+						continue // when GR, is normal to observe stale routes
+					}
+					Expect(err).NotTo(HaveOccurred(), "downtime was observed")
+				}
+
+				pods, err := metallb.SpeakerPods(cs)
+				if err != nil {
+					return err
+				}
+
+				for _, p := range pods {
+					if !k8s.PodIsReady(p) {
+						return fmt.Errorf("speaker pods are not ready")
+					}
+				}
+
+				return nil
+			}, 2*time.Minute, time.Second).ShouldNot(HaveOccurred(), "no downtime until speakers are ready")
+
+			for _, c := range FRRContainers {
+				validateService(svc, allNodes.Items, c)
+			}
+		}
+
+		ginkgo.Context("and when GR enabled", func() {
+
+			assertDuringSpeakerRestartWithGR := func(pairingIPFamily ipfamily.Family, poolAddresses []string, tweak testservice.Tweak) {
+				assertDuringSpeakerRestart(true, pairingIPFamily, poolAddresses, tweak)
```

Comment:
nit: let's define a constant "EnableGracefulRestart = true" and let's use that instead of true / false (also when checking)

Reply:
done